### PR TITLE
sd_ass: don't reconfigure ass on every frame

### PR DESCRIPTION
**Pull request by Dudemanguy**
dbc5d7b7db70f898dcd5cbbba76ffed38e5371ee seems to have originally introduced this behavior. At the time, wm4 simply reconfigured ass on every frame in order to accommodate runtime changes in sub options. This certainly works, but these libass API calls are not free and there is at least one known performance regression due to a change in libass*. Regardless of whether or not the libass change is good/bad, there is no need for mpv to constantly reconfigure this. When wm4 made that commit, there was no notification mechanism for options changing that could easily be used so he didn't really have any other choice. But it's nearly 10 years later now and internally we have all the necessary pieces to only configure ass again when we need to: on option changes or resizes. So go ahead and implement that in this commit which simply uses the already existing SD_CTRL_UPDATE_OPTS and compares osd_res sizes to determine whether or not an ass configure is needed.

*: https://github.com/libass/libass/issues/698

A couple of related issues: #11612 and #11511. This should help in theory but the sample in the first issue always drops a tons of frames for me no matter what (haven't tried the samples in the second issue yet).

### CODE REVIEW

File: sub/sd_ass.c [hunk 1]
```diff
@@ -831,6 +840,7 @@ static int control(struct sd *sd, enum sd_ctrl cmd, void *arg)
             assobjects_destroy(sd);
             assobjects_init(sd);
         }
+        ctx->ass_configured = false; // ass always needs to be reconfigured
```
Comment (sfan5) [inline sub/sd_ass.c:843]: i saw the `reset()` above and am wondering: does it need to be there too?

Comment (Dudemanguy) [reply sub/sd_ass.c:843]: I don't believe so. The `reset` function seems to trigger when sub tracks are changed/reloaded/etc. (happens up in the core of mpv). These conditions don't require reconfiguring ass.

**Review by sfan5 — APPROVED** [review]

**Comment by llyyr** [comment]
Probably worth documenting that this change may not necessarily fix those two issues, and they might not even be mpv's issues in the first place. After some benchmarking, all that this change does is make [1a599b1](https://github.com/libass/libass/commit/1a599b1db748dfc07df98cd37d8603edb87da2fd) not regress performance in mpv. If some subtitles already had poor performance before that upstream commit, as is the case for those two issues, then they'll likely stay the same or only have marginal performance benefits.

This should help with the libass issue linked, however.

**Comment by Dudemanguy** [comment]
Both reporters are windows users that presumably are using Shinchiro's builds after that libass commit, so they *do* get worse performance than anyone using a point release of libass. As for solving their problem, yeah probably not (although one guy claims he was actually able to play it before), but at least mpv shouldn't be actively making it worse anymore with this.

**Pull request merged**